File: symmetric-encryption/js/decrypt.js
#!/usr/bin/env node

const fs = require('fs');
const crypto = require('crypto');

// Read the hex-encoded secret key from the first command-line argument.
const secret = Buffer.from(process.argv[2], 'hex');

// Read the data to be decrypted from stdin.
const encryptedData = fs.readFileSync(0, 'utf-8').trim();

// Extract and decode the provided HMAC digest.
let [ data, providedDigest ] = encryptedData.split('--');
providedDigest = Buffer.from(providedDigest, 'hex');

// Calculate the digest ourselves.
const calculatedDigest = crypto.createHmac('sha1', secret).update(data).digest();

// Compare the provided and the calculated digests securely to ensure the
// encrypted data is authentic.
if (!crypto.timingSafeEqual(providedDigest, calculatedDigest)) {
	throw new Error('Invalid digest.');
}

// Split the encrypted data from the IV.
let [ ciphertext, iv ] = Buffer.from(data, 'base64').toString().split('--');
iv = Buffer.from(iv, 'base64');

// Decrypt the data using the secret and IV.
const decipher = crypto.createDecipheriv('aes256', secret, iv);
const plaintext = decipher.update(ciphertext, 'base64', 'utf-8') + decipher.final('utf-8');

console.log(plaintext);

File: symmetric-encryption/python/decrypt.py
#!/usr/bin/env python3

import sys
import hmac
import base64
import binascii

from Cryptodome.Cipher import AES
from Cryptodome.Util.Padding import unpad

if __name__ == '__main__':
	# Read the hex-encoded secret key from the first command-line argument.
	key = binascii.unhexlify(sys.argv[1])

	# Read the data to be decrypted from stdin.
	auth_data = sys.stdin.buffer.read().strip()

	# Extract and decode the provided HMAC digest.
	data, provided_hmac = auth_data.split(b'--')

	# Calculate the digest ourselves.
	calculated_hmac = binascii.hexlify(hmac.digest(key, data, 'sha1'))

	# Compare the provided and the calculated digests securely to ensure the
	# encrypted data is authentic.
	hmac_check = hmac.compare_digest(provided_hmac, calculated_hmac)
	if not hmac_check:
		raise Exception('Failed HMAC check')

	# Split the encrypted data from the IV.
	ciphertext, iv = base64.b64decode(data).split(b'--')

	# Decrypt the data using the secret and IV.
	cipher = AES.new(key, AES.MODE_CBC, iv=base64.b64decode(iv))
	plaintext = unpad(cipher.decrypt(base64.b64decode(ciphertext)), AES.block_size)

	print(plaintext)
